numpy/core/shape_base.py

```diff
@@ -275,3 +278,62 @@ def hstack(tup):
         return _nx.concatenate(arrs, 0)
     else:
         return _nx.concatenate(arrs, 1)
+
+def stack(arrays, axis=0):
+    """
+    Join a sequence of arrays along a new axis.
+
+    .. versionadded:: 1.10.0
+
+    Parameters
+    ----------
+    arrays : sequence of ndarrays
+        Each array must have the same shape.
+    axis : int, optional
+        The axis along which the arrays will be stacked.
+
+    Returns
+    -------
+    stacked : ndarray
+        The stacked array has one more dimension than the input arrays.
+
+    See Also
+    --------
+    concatenate : Join a sequence of arrays along an existing axis.
+    split : Split array into a list of multiple sub-arrays of equal size.
+
+    Examples
+    --------
+    >>> arrays = [np.random.randn(3, 4) for _ in range(10)]
+    >>> np.stack(arrays, axis=0).shape
+    (10, 3, 4)
+
+    >>> np.stack(arrays, axis=1).shape
+    (3, 10, 4)
+
+    >>> np.stack(arrays, axis=2).shape
+    (3, 4, 10)
+
+    >>> a = np.array([1, 2, 3])
+    >>> b = np.array([2, 3, 4])
+    >>> np.stack((a, b))
+    array([[1, 2, 3],
+           [2, 3, 4]])
+
+    >>> np.stack((a, b), axis=-1)
+    array([[1, 2],
+           [2, 3],
+           [3, 4]])
+
+    """
+    arrays = [_nx.asarray(arr) for arr in arrays]
+    if not arrays:
+        raise ValueError('need at least one array to stack')
+    result_ndim = arrays[0].ndim + 1
+    if not -result_ndim <= axis < result_ndim:
+        raise IndexError('axis %r out of bounds [-%r, %r)'
+                         % (axis, result_ndim, result_ndim))
+    if axis < 0:
+        axis += result_ndim
+    sl = (slice(None),) * axis + (_nx.newaxis,)
+    return _nx.concatenate([arr[sl] for arr in arrays], axis=axis)
```

Comment:
@shoyer - actually, looking better both your method and mine can lead to unexpected errors when a later array has fewer dimensions than the first one, like::
```
np.stack([np.arange(9.).reshape(3,3), np.arange(3)], axis=2)
```
At least, I would think you would get `sl=[:,:,np.newaxis]` in which case you get an `IndexError: too many indices` on the second array (rather than the much more informative `ValueError` that `concatenate` would throw).

Though it would perhaps be needlessly expensive to test for this beforehand. Maybe just
```
try:
    return _nx.concatenate(...)
except IndexError:
    raise ValueError(<like concatenate>)
```